content/advent-2018/go-in-the-browser.md

```diff
@@ -0,0 +1,175 @@
+
+```go
+s.onImgLoadCb = js.NewCallback(func(args []js.Value) {
+	reader := bytes.NewReader(s.inBuf)
+	var err error
+	s.sourceImg, _, err = image.Decode(reader)
+	if err != nil {
+		s.log(err.Error())
+		return
+	}
+	// Now the sourceImg is an image.Image with which we are free to do anything !
+})
+```
+
+Then we take user values from any of the effect sliders, and manipulate the image. We use the awesome [bild](https://github.com/anthonynsimon/bild) library for that. Here is a small snippet of the contrast callback -
+
+```go
+s.contrastCb = js.NewEventCallback(js.PreventDefault, func(ev js.Value) {
+	delta := ev.Get("target").Get("valueAsNumber").Float()
+	res := adjust.Contrast(s.sourceImg, delta)
+})
+```
+
+After this, we encode the target image to jpeg and send it back to the browser. Here is the full app in action -
+
+We load the image:
+
+![initial](/postimages/advent-2018/go-in-the-browser/initial.png)
+
+Change contrast:
+
+![contrast](/postimages/advent-2018/go-in-the-browser/contrast.png)
+
+Change hue:
+
+![hue](/postimages/advent-2018/go-in-the-browser/hue.png)
+
+Awesome, we are able to natively manipulate images in the browser without writing a single line of Javascript ! The source code can be found [here](https://github.com/agnivade/shimmer).
+
+Note that all this is being done natively in the browser itself. There are no Flash plugins, Java Applets or Silverlight magic happening here. WebAssembly is supported natively in the browser out of the box.
+
+## Final words
+
+Some of my closing remarks:
+
+- Since Go is a garbage collected language, the entire runtime is shipped inside the wasm binary. Hence it is common for binaries to have large sizes in the order of MBs. This is still a sore point compared to other languages like C/Rust; because shipping MBs of data to the browser is not ideal. However, if the wasm spec supports GC by itself, then this can change.
+- Wasm support in Go is officially experimental. The `syscall/js` API itself is in flux and might change in future. If you see a bug, please feel free to file an issue at our [issue tracker](https://github.com/golang/go/issues).
+- Like all technologies, WebAssembly is not a silver bullet. Sometimes, simple JS is faster and easier to write. However, the wasm spec itself is very much in development, and there are more features coming soon. Thread support is one such feature.
+
+Hopefully, this post showed some of the cool aspects of WebAssembly and how you can write a fully-functioning webapp using Go. Do try it out, and file issues if you see a bug. If you need any help, feel free to drop in to the [#webassembly](https://gophers.slack.com/) channel.
```

Comment:
You've mentioned an issue in the paragraph above. I think you can skip repeating it here.

```suggestion
Hopefully, this post showed some of the cool aspects of WebAssembly and how you can write a fully-functioning web app using Go. Go try it out. If you need any help, feel free to drop in the [#webassembly](https://gophers.slack.com/) channel.
```

Reply:
Removed the issues part. I think `to` is the appropriate preposition here. I added that deliberately.